### ISSUE TITLE
Cannot open and close the connection multiple times

### ISSUE THREAD

**Issue by adaxi** [memphisdev/memphis.js#187]
**Bug description**
When opening, then closing, then reopening the connection, uncaught exceptions are thrown:

```
node:internal/process/promises:288
            triggerUncaughtException(err, true /* fromPromise */);
            ^

memphis: CONNECTION_CLOSED
    at memphis.errorForCode (/home/gbo/Workspaces/@escaux/message-broker/node_modules/nats/lib/nats-base-client/error.js:98:16)
    at NatsConnectionImpl._check (/home/gbo/Workspaces/@escaux/message-broker/node_modules/nats/lib/nats-base-client/nats.js:94:37)
    at NatsConnectionImpl.subscribe (/home/gbo/Workspaces/@escaux/message-broker/node_modules/nats/lib/nats-base-client/nats.js:116:14)
    at JetStreamClientImpl.fetch (/home/gbo/Workspaces/@escaux/message-broker/node_modules/nats/lib/nats-base-client/jsclient.js:164:29)
    at Consumer.fetch (/home/gbo/Workspaces/@escaux/message-broker/node_modules/memphis-dev/lib/consumer.js:107:66)
    at Timeout.fetchAndHandleMessages [as _onTimeout] (/home/gbo/Workspaces/@escaux/message-broker/node_modules/memphis-dev/lib/consumer.js:59:49)
    at listOnTimeout (node:internal/timers:569:17)
    at process.processTimers (node:internal/timers:512:7)
Emitted 'error' event at:
    at Timeout.fetchAndHandleMessages [as _onTimeout] (/home/gbo/Workspaces/@escaux/message-broker/node_modules/memphis-dev/lib/consumer.js:63:39) {
  name: 'MemphisError',
  code: 'CONNECTION_CLOSED',
  chainedError: undefined
}

Node.js v18.17.1
```

**To Reproduce**

Execute the following script:

```js
import { memphis } from 'memphis-dev'


const test = async () => {
  const memphisConnection1 = await memphis.connect({
    host: 'localhost',
    port: 6666,
    username: 'root',
    password: 'memphis'
  })

  const consumer1 = await memphisConnection1.consumer({
    consumerName: 'DefaultConsumer',
    consumerGroup: 'DefaultConsumerGroup',
    stationName: 'test.topic'
  })

  consumer1.on('message', () => console.log('consumer 1 got a message'))


  await memphisConnection1.close()

  const memphisConnection2 = await memphis.connect({
    host: 'localhost',
    port: 6666,
    username: 'root',
    password: 'memphis'
  })

  const consumer2 = await memphisConnection2.consumer({
    consumerName: 'DefaultConsumer',
    consumerGroup: 'DefaultConsumerGroup',
    stationName: 'test.topic'
  })

  consumer2.on('message', () => console.log('consumer 2 got a message'))


  await memphisConnection2.close()
}

test()
```

**Expected behavior**
I would expect to be able to open then close multiple times.

**Additional context**
I am using image: memphisos/memphis:1.2.0 and SDK v1.1.2

**Comment by yanivbh1** [comment]
Hey @adaxi ,
Thanks for raising it up.
Adding @shay23b and @shohamroditimemphis

**Comment by obumnwabude** [comment]
> I would expect to be able to open then close multiple times.

Hi @adaxi, please what's the use case of closing and restarting the Memphis connection multiple times?

I'm not so sure about the following, but I can say that by design, the connection is not intended to be re-opened after being closed. 

I'm supposing that the close method is available in case a fatal error occurs in the code and it's best to stop any activity with the broker. In fact, the main use case of this method is if the broker didn't successfully connect at the start, then you call .close() in the catch block to save resources kind of.

Chances are if something should make you close the entire Memphis connection, then that thing will equally warrant a server restart. 

Like I said I'm not sure, this is my opinion. Or rather,

What is the use case of the multiple opens and closes?

**Comment by obumnwabude** [comment]
@yanivbh1 while pending @adaxi 's response,

I suppose a temporary solution could be to re-assign the default exported `memphis` instance on close of the connection.

So instead of 

https://github.com/memphisdev/memphis.js/blob/0b9fdeb2334d57304b32cbbd9837579b9282d439/src/memphis.ts#L1239C1-L1239C1

We could have 

```js
// somewhere at the top of the memphis.ts file
var memphis = new Memphis();

// in the close function, after the entire logic, we will re-assign to a new instance
memphis = new Memphis();

// then at the end of the file we just export this instance
// that is instead of a 'const' version of it 
export memphis;
```

This way after closing, a new re-connection in importing codebases will technically be a new Memphis instance (unknown to the user) which doesn't break code. Just a suggestion though. 

Though I'm the same person proposing this temporary solution, it still doesn't look good to me. 

This re-assignment means that there will be a need to recreate producers and consumers as all the existing ones would have been discarded in the close() function. In other words, the importing codebase will have to redeclare them on reconnecting Memphis. 

And this comes back to my previous comment that by design the Memphis instance should not be re-opened after close. You can see it just as your NodeJS server should keep listening throughout its life. 

This is also why the memphis connection is exported as a singleton instance. Like we are exporting an instance of the Memphis class and not the Memphis class itself. We don't expect any initialization of the Memphis class in importing libraries.

**Comment by obumnwabude** [comment]
https://github.com/memphisdev/memphis.js/blob/0b9fdeb2334d57304b32cbbd9837579b9282d439/src/memphis.ts#L1239C1-L1239C1

**Comment by adaxi** [comment]
Perhaps I started the bug report by the wrong end.

In some cases we are experiencing `Produce operation has failed, please check whether Station/Producer still exist` when producing a message. This happens when we recreate a new instance of memphis, but do not to restart the clients. 

I expect the SDK to recreate the producer if needed (we are using the `produce` method). I attempted to restart a new connection, and then encountered this bug.

**Comment by adaxi** [comment]
It also breaks our test suite. We have a bunch of tests where for each test we start and stop the connection.

**Comment by shohamroditimemphis** [comment]
I will take  a look and will respond ASAP @adaxi

**Comment by idanasulinmemphis** [comment]
This will be fixed in 1.2.0, closing this one
